packages/superset-ui-query/src/types/QueryFormData.ts

```diff
@@ -1,18 +1,17 @@
 /* eslint camelcase: 0 */
 // FormData uses snake_cased keys.
-import { MetricKey, AdhocMetric } from './Metric';
+import { AdhocMetric } from './Metric';
 import { TimeRange } from './Time';
 import { AdhocFilter } from './Filter';
 
 export type QueryFormDataMetric = string | AdhocMetric;
-
-// Define mapped type separately to work around a limitation of TypeScript
-// https://github.com/Microsoft/TypeScript/issues/13573
-// The Metrics in formData is either a string or a proper metric. It will be
-// unified into a proper Metric type during buildQuery (see `/query/Metrics.ts`).
-export type QueryFormDataMetrics = Partial<
-  Record<MetricKey, QueryFormDataMetric | QueryFormDataMetric[]>
->;
+export type ResidualFormData = {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  [key: string]: any;
```

Comment:
ditto on `unknown`

Reply:
Actually, are the items in ResidualFormData a `QueryFormDataMetric`? `buildGroupedControlData` seems to imply that.

Reply:
Yes, I was also puzzled by this. `ResidualFormData` (in retrospect should be `QueryFormResidualData`) is currently expected to contain either column names (`string`) or metrics (`QueryFormDataMetric`  which is the same as `string | AdhocMetric`). When these are combined, the union is reduced to `string | AdhocMetric` which is the same as `QueryFormDataMetric`. I'm tempted to remove `QueryFormDataMetric` and just pass around `QueryFormResidualData`, which can later be split into `QueryFormDataMetric` and `QueryFormDataColumn` when that need arises (Down the line we'll need to add `AdhocColumn` which is a column with a SQL expression). I will actually do that, we'll see how it looks.